File: YourBible/plugins/com-shortsands-audioplayer/src/ios/AudioPlayer/AudioTOCBible.swift
//
//  AudioTOCBible.swift
//  AudioPlayer
//
//  Created by Gary Griswold on 8/7/17.
//  Copyright © 2017 ShortSands. All rights reserved.
//

import Foundation
#if USE_FRAMEWORK
    import AWS
#endif

class AudioTOCBible {
    
    let textVersion: String
    let silLang: String
    let database = AudioSqlite3()
    var oldTestament: AudioTOCTestament?
    var newTestament: AudioTOCTestament?
    
    init(versionCode: String, silLang: String) {
        self.textVersion = versionCode
        self.silLang = silLang
        self.oldTestament = nil
        self.newTestament = nil
    }
    
    deinit {
        print("***** Deinit AudioMetaDataReader *****")
    }

    func read(complete: @escaping (_ oldTest:AudioTOCTestament?, _ newTest:AudioTOCTestament?) -> Void) {
        let query = "SELECT a.damId, a.collectionCode, a.mediaType, a.dbpLanguageCode, a.dbpVersionCode" +
                " FROM audio a, audioVersion v" +
                " WHERE a.dbpLanguageCode = v.dbpLanguageCode" +
                " AND a.dbpVersionCode = v.dbpVersionCode" +
                " AND v.versionCode = ?" +
                " ORDER BY mediaType ASC, collectionCode ASC"
                // mediaType sequence Drama, NonDrama
                // collectionCode sequence NT, ON, OT
        do {
            try self.database.open(dbPath: "Versions.db", copyIfAbsent: true)
            defer { self.database.close() }
            try self.database.queryV1(sql: query, values: [self.textVersion], complete: { resultSet in
                print("LENGTH \(resultSet.count)")
                var oldTestRow: [String?]? = nil
                var newTestRow: [String?]? = nil
                for row in resultSet {
                    // Because of the sort sequence, the following logic prefers Drama over Non-Drama
                    // Because of the sequenc of IF's, it prefers OT and NT over ON
                    let collectionCode = row[1]!
                    if newTestRow == nil && collectionCode == "NT" {
                        newTestRow = row
                    }
                    if newTestRow == nil && collectionCode == "ON" {
                        newTestRow = row
                    }
                    if oldTestRow == nil && collectionCode == "OT" {
                        oldTestRow = row
                    }
                    if oldTestRow == nil && collectionCode == "ON" {
                        oldTestRow = row
                    }
                }
                if let oldRow = oldTestRow {
                    self.oldTestament = AudioTOCTestament(bible: self, database: self.database,
                                                          mediaSource: "FCBH", dbRow: oldRow)
                }
                if let newRow = newTestRow {
                    self.newTestament = AudioTOCTestament(bible: self, database: self.database,
                                                          mediaSource: "FCBH", dbRow: newRow)
                }
                self.readBookNames()
                complete(self.oldTestament, self.newTestament)
            })
        } catch let err {
            print("ERROR \(AudioSqlite3.errorDescription(error: err))")
            complete(nil, nil)
        }
    }
    /**
    * This function will only return results after read has been called.
    */
    func findBook(bookId: String) -> AudioTOCBook? {
        var result: AudioTOCBook? = nil
        if let oldTest = self.oldTestament {
            result = oldTest.booksById[bookId]
        }
        if result == nil {
            if let newTest = self.newTestament {
                result = newTest.booksById[bookId]
            }
        }
        return result
    }
    
    func readVerseAudio(damid: String, bookId: String, chapter: Int,
                        complete: @escaping (_ audioVerse: AudioTOCChapter?) -> Void) {
        var metaDataVerse: AudioTOCChapter? = nil
        let query = "SELECT versePositions FROM AudioChapter WHERE damId = ? AND bookId = ? AND chapter = ?"
        do {
            try self.database.open(dbPath: "Versions.db", copyIfAbsent: true)
            defer { self.database.close() }
            try self.database.queryV1(sql: query, values: [damid, bookId, String(chapter)], complete: { resultSet in
                print("LENGTH \(resultSet.count)")
                if resultSet.count > 0 {
                    let row = resultSet[0]
                    if let verses = row[0] {
                        metaDataVerse = AudioTOCChapter(json: verses)
                    }
                }
                complete(metaDataVerse)
            })
        } catch let err {
            print("ERROR \(AudioSqlite3.errorDescription(error: err))")
            complete(nil)
        }
    }
    
    private func readBookNames() {
        let query = "SELECT code, heading FROM tableContents"
        do {
            let dbName = self.textVersion + ".db"
            try self.database.open(dbPath: dbName, copyIfAbsent: true)
            defer { self.database.close() }
            try self.database.queryV1(sql: query, values: nil, complete: { resultSet in
                for row in resultSet {
                    let bookId = row[0]!
                    if let oldTest = self.oldTestament?.booksById[bookId] {
                        oldTest.bookName = row[1]!
                    } else if let newTest = self.newTestament?.booksById[bookId] {
                        newTest.bookName = row[1]!
                    }
                }
            })
        } catch let err {
            print("ERROR \(AudioSqlite3.errorDescription(error: err))")
        }
    }
    
    /*
    * Deprecated. This was used to parse verse positions as a dictionary
    private func parseJsonDictionary(json: String) -> NSDictionary? {
        var result: NSDictionary? = nil
        if let data = json.data(using: .utf8) {
            do {
                result = try JSONSerialization.jsonObject(with: data, options:.allowFragments) as? NSDictionary
            } catch let jsonError {
                print("Error parsing Meta Data json \(jsonError)")
            }
        }
        return result
    }
    */
}

File: Plugins/AudioPlayer/src/android/app/src/main/java/com/shortsands/audioplayer/AudioTOCBible.java
package com.shortsands.audioplayer;

/**
 * Created by garygriswold on 8/30/17.
 */
import android.content.Context;
import android.database.Cursor;
import android.util.Log;

class AudioTOCBible {

    private static final String TAG = "AudioTOCBible";

    private Context context;
    // which of these should be private, none of them are in ios
    String textVersion;
    String silLang;
    String mediaSource;
    AudioTOCTestament oldTestament;
    AudioTOCTestament newTestament;
    AudioSqlite3 database;

    AudioTOCBible(Context context, String versionCode, String silLang) {
        this.context = context;
        this.textVersion = versionCode;
        this.silLang = silLang;
        this.mediaSource = "FCBH";
        this.oldTestament = null;
        this.newTestament = null;
        this.database = new AudioSqlite3(context);
        try {
            this.database.open("Versions.db", true);
        } catch (Exception err) {
            Log.d(TAG,"ERROR " + AudioSqlite3.errorDescription(err));
        }
    }

    //deinit {
    //    print("***** Deinit AudioMetaDataReader *****")
    //}

    void read() {
        String query = "SELECT a.damId, a.collectionCode, a.mediaType, a.dbpLanguageCode, a.dbpVersionCode" +
                " FROM audio a, audioVersion v" +
                " WHERE a.dbpLanguageCode = v.dbpLanguageCode" +
                " AND a.dbpVersionCode = v.dbpVersionCode" +
                " AND v.versionCode = ?" +
                " ORDER BY mediaType ASC, collectionCode ASC";
        // mediaType sequence Drama, NonDrama
        // collectionCode sequence NT, ON, OT
        try {
            String[] values = new String[1];
            values[0] = this.textVersion;
            Cursor cursor = this.database.queryV2(query, values);
            Log.d(TAG, "LENGTH " + cursor.getCount());
            while (cursor.moveToNext()) {
                // Because of the sort sequence, the following logic prefers Drama over Non-Drama
                // Because of the sequence of IF's, it prefers OT and NT over ON
                String collectionCode = cursor.getString(1);
                if (this.newTestament == null && collectionCode.equals("NT")) {
                    this.newTestament = new AudioTOCTestament(this, this.database, cursor);
                }
                if (this.oldTestament == null && collectionCode.equals("OT")) {
                    this.oldTestament = new AudioTOCTestament(this, this.database, cursor);
                }
            }
            this.readBookNames(); // This is not necessary on Android
        } catch (Exception err) {
            Log.d(TAG,"ERROR " + AudioSqlite3.errorDescription(err));
        }
    }
    /**
     * This function will only return results after read has been called.
     */
    AudioTOCBook findBook(String bookId) {
        AudioTOCBook result = null;
        if (this.oldTestament != null) {
            result = this.oldTestament.booksById.get(bookId);
        }
        if (result == null) {
            if (this.newTestament != null) {
                result = this.newTestament.booksById.get(bookId);
            }
        }
        return result;
    }

    AudioTOCChapter readVerseAudio(String damid, String bookId, int chapter) {
        AudioTOCChapter tocChapter = null;
        String query = "SELECT versePositions FROM AudioChapter WHERE damId = ? AND bookId = ? AND chapter = ?";
        try {
            String[] values = new String[3];
            values[0] = damid;
            values[1] = bookId;
            values[2] = String.valueOf(chapter);
            Cursor cursor = this.database.queryV2(query, values);
            while(cursor.moveToNext()) {
                String positions = cursor.getString(0);
                if (positions != null) {
                    tocChapter = new AudioTOCChapter(positions);
                }
            }
        } catch (Exception err) {
            Log.d(TAG,"ERROR " + AudioSqlite3.errorDescription(err));
        } finally {
            return(tocChapter);
        }
    }

    private void readBookNames() {
        String query = "SELECT code, heading FROM tableContents";
        String[] values = new String[0];
        AudioSqlite3 db = new AudioSqlite3(this.context);
        try {
            String dbName = this.textVersion + ".db";
            db.open(dbName, true);
            String[][] resultSet = db.queryV1(query, values);
            for (int i=0; i<resultSet.length; i++) {
                String[] row = resultSet[i];
                String bookId = row[0];
                if (this.oldTestament != null && this.oldTestament.booksById.containsKey(bookId)) {
                    this.oldTestament.booksById.get(bookId).bookName = row[1];
                } else if (this.newTestament != null && this.newTestament.booksById.containsKey(bookId)) {
                    this.newTestament.booksById.get(bookId).bookName = row[1];
                }
            }
        } catch (Exception err) {
            Log.d(TAG, "ERROR: " + AudioSqlite3.errorDescription(err));
        } finally {
            db.close();
        }
    }
}
